Gated infrastructure deploy template (#21)
UKHO/devops-pipelinetemplates: gated-infrastructure-deploy/README.md

# Template for Gated Infrastructure Deploy

> This template is designed to allow you to deploy terraform resources with a manually verification if any resources are getting destroyed. This manual verification will only trigger if there is indication that resources will be destroyed. If triggered, then the job will fail after a timeout has been reached. If there is no indication that resources wil be destroyed, the template will carry on and deploy the resources.

Reviewer: ```suggestion
This template is designed to allow you to deploy resources via Terraform with a manual verification step if the Terraform `plan` indicates that any resources will be destroyed. When no resources are flagged for destruction then the pipeline will automatically continue to Terraform `apply`. If the verification job is triggered, then you will need to approve in Azure DevOps in order to move on to the `apply` step. The job will fail if you either reject or do not approve within the timeout period.
```

This template makes use of `terraform plan` and `terraform apply`.

**The plan cannot foresee every circumstance and scenario compared to the apply.**

## Usage

1. Reference this repository in your root `azure-pipelines.yml` file.

    ```yaml
    resources:
      repositories:
        - repository: UKHOTemplates
          type: github
          endpoint: UKHO
          name: UKHO/devops-pipelinetemplates
          ref: refs/heads/gated-infrastructure-deploy-template
    ```

> 2. Add this template block to your pipeline and fill out the values that relate to you deployment. The template content is a series of `jobs`, therefore they need to be rooted under a `stage`.

Reviewer: ```suggestion
2. Add this template block to your pipeline and fill out the values that relate to your deployment. The template content is a series of `jobs`, therefore they need to be rooted under a `stage`.
```

    | TerraformArtifactConfigRelativePath | Yes | string | Relative path to the .tf files inside your artifact |
    | VariablesTemplateRelativePath | Yes | string | Relative path to a YAML template containing a variables expression |
    | TerraformVariableMappings | Yes | object | A key/value map of Terraform variables to be injected into the PowerShell runtime environment for Terraform to use |
    | TerraformOutputVariables| No | object | An array of Terraform output variables to be retrieved after the Terraform apply has completed |

    **Notes**:

    - *TerraformWorkspace*: This value will be appended onto the blob name in the form `[TFStateBlobName]:[TerraformWorkspace]`, e.g. `terraform.deployment.tfplan:dev`   

    - *TerraformArtifact*: The files in the artifact will be used without modifying their contents. Your pipeline will need to include jobs to make any required modfications to your supporting files e.g. inserting configuration values, before publishing the artifact.

    - *TerraformArtifactConfigRelativePath*: Inside the template, the full path to the configuration will be `$(Pipeline.Workspace)/[TerraformArtifact][TerraformArtifactConfigRelativePath]`.
      

>       Example 1: The artifact is named 'tfartifact' and the .tf files are in the root. `TerraformArtifactConfigRelativePath` would be `/` and the concatenated path would be `$(Pipeline.Workspace)/terraformartifact/`.

Author: ```suggestion
      Example 1: The artifact is named 'tfartifact' and the .tf files are in the root. `TerraformArtifactConfigRelativePath` would be `/` and the concatenated path would be `$(Pipeline.Workspace)/tfartifact/`.
```

Adding example name into the concatenated path.

>       ```
>       +-- terraform
>       |   +-- azure.tf
>       |   +-- main.tf
>       |   +-- variables.tf
>       |   +-- output.tf
>       ```

Author: ```suggestion
      ```
      +-- tfartifact
      |   +-- azure.tf
      |   +-- main.tf
      |   +-- variables.tf
      |   +-- output.tf
      ```
```

Aligning with the example.

Reviewer: Difference between folder structure and artifact unclear, resolved by commit 45fe820

>       ```
>       +-- terraform
>       |   +-- azure.tf
>       |   +-- main.tf
>       |   +-- variables.tf
>       |   +-- output.tf
>       ```

Author: ```suggestion
      ```
      +-- tfartifact
      |   +-- azure.tf
      |   +-- main.tf
      |   +-- variables.tf
      |   +-- output.tf
      ```
```

Aligning with the example.

      Example 2: The artifact is named 'buildartifact' and the .tf files are in a subfolder. `TerraformArtifactConfigRelativePath` would be `terraform/` and the concatenated path would be `$(Pipeline.Workspace)/buildartifact/terraform/`.

>       ```
>       +-- src
>       +-- terraform
>       |   +-- azure.tf
>       |   +-- main.tf
>       |   +-- variables.tf
>       |   +-- output.tf
>       ```

Author: ```suggestion
      ```
      +-- buildartifact
      |   +-- terraform
      |       +-- azure.tf
      |       +-- main.tf
      |       +-- variables.tf
      |       +-- output.tf
      ```
```
Updating to be inline with the example.

Reviewer: Difference between folder structure and artifact unclear, resolved by commit 45fe820
